src/applicationcomposer/messageHandlers/deployMessageHandler.ts

```diff
@@ -7,6 +7,45 @@ import vscode from 'vscode'
 import { Command, DeployRequestMessage, DeployResponseMessage, MessageType, WebviewContext } from '../types'
 import { SamSyncResult } from '../../shared/sam/sync'
 import { telemetry } from '../../shared/telemetry/telemetry'
+import { Auth } from '../../auth/auth'
+import { promptAndUseConnection } from '../../auth/utils'
+import { StatefulConnection } from '../../auth/connection'
+import * as nls from 'vscode-nls'
+import { showMessageWithUrl } from '../../shared/utilities/messages'
+import { credentialHelpUrl } from '../../shared/constants'
+
+const localize = nls.loadMessageBundle()
+
+async function promptReauth(connection?: StatefulConnection) {
+    let errorMessage = localize(
+        'aws.applicationComposer.deploy.authModal.message',
+        'Syncing requires authentication with IAM credentials.'
+    )
+    if (connection?.state === 'valid') {
+        errorMessage =
+            localize(
+                'aws.applicationComposer.deploy.authModal.invalidAuth',
+                'Authentication through Builder ID or IAM Identity Center detected. '
+            ) + errorMessage
+    }
+    const acceptMessage = localize(
+        'aws.applicationComposer.deploy.authModal.accept',
+        'Authenticate with IAM credentials'
+    )
+    const docMessage = localize('aws.applicationComposer.deploy.authModal.docLink', 'Open documentation')
+    const modalResponse = await showMessageWithUrl(
+        errorMessage,
+        credentialHelpUrl,
+        docMessage,
+        'info',
+        [acceptMessage],
+        true
+    )
+    if (modalResponse !== acceptMessage) {
+        return
+    }
+    await promptAndUseConnection(Auth.instance, 'iam-only')
```

Comment:
Could we make this change directly in https://github.com/aws/aws-toolkit-vscode/blob/ea875fc9af230a292e7a65b2a0d5f034a9314be2/src/shared/sam/sync.ts#L547-L548 instead of only for AppComposer?

Reply:
Yes, just moved it inside the SAM sync.